### DISCUSSION TITLE
Render to memory without copy

### DISCUSSION THREAD

**Discussion by elbaro** [gfx-rs/wgpu#4292] [Q&A] [answered]
I am following [Wgpu without a window](https://sotrh.github.io/learn-wgpu/showcase/windowless/#so-what-do-we-need-to-do)  with the code:

```rs
let buffer_slice = output_buffer.slice(..);
let (tx, rx) = futures_intrusive::channel::shared::oneshot_channel();
buffer_slice.map_async(wgpu::MapMode::Read, move |result| {
    tx.send(result).unwrap();
});
self.device.poll(wgpu::Maintain::Wait);
pollster::block_on(rx.receive()).unwrap().unwrap();

let data = buffer_slice.get_mapped_range();
use image::{ImageBuffer, Rgba};
let buffer = ImageBuffer::<Rgba<u8>, _>::from_raw(
    self.size.width,
    self.size.height,
    data,
)
.unwrap();
buffer.save("image.png").unwrap();
output_buffer.unmap();
```

Now I want to abstract this to `fn render(..) -> Vec<u8>` but BufferView<'a> is bound on BufferSlice, and on Buffer. Can wgpu directly render to a provided host memory, or can the user take the ownership of the host Buffer data and detach it from wgpu context?

Copying the data from gpu to a host memory (buffer) and then copying again to Vec<u8> sounds unnecessary.

**Comment by SludgePhD** [comment, marked as answer]
`Buffer` is located in GPU memory, and `BufferView` is a view into that GPU memory (mapped into your programs address space, so that it can be treated as normal memory). The copy from that to the `Vec<u8>` would only copy the data to CPU memory once.
